pkg/checks/dns/dns.go

```diff
@@ -0,0 +1,253 @@
+			res := d.check(ctx)
+			errval := ""
+			r := types.Result{
+				Data:      res,
+				Err:       errval,
+				Timestamp: time.Now(),
+			}
+
+			d.CResult <- r
+			log.Debug("Successfully finished dns check run")
+		}
+	}
+}
+
+func (d *DNS) Startup(ctx context.Context, cResult chan<- types.Result) error {
+	log := logger.FromContext(ctx)
+	log.Debug("Initializing dns check")
+
+	d.CResult = cResult
+	return nil
+}
+
+func (d *DNS) Shutdown(_ context.Context) error {
+	d.Done <- true
+	close(d.Done)
+
+	return nil
+}
+
+func (d *DNS) SetConfig(ctx context.Context, conf any) error {
+	log := logger.FromContext(ctx)
+
+	c, err := helper.Decode[config](conf)
+	if err != nil {
+		log.Error("Failed to decode dns config", "error", err)
+		return errors.ErrInvalidConfig
+	}
+
+	d.Mu.Lock()
+	defer d.Mu.Unlock()
+	d.config = c
+
+	return nil
+}
+
+// Schema provides the schema of the data that will be provided
+// by the dns check
+func (d *DNS) Schema() (*openapi3.SchemaRef, error) {
+	return checks.OpenapiFromPerfData(make(map[string]Result))
+}
+
+// RegisterHandler registers a server handler
+func (d *DNS) RegisterHandler(_ context.Context, router *api.RoutingTree) {
+	router.Add(http.MethodGet, route, d.Handler)
+}
+
+// DeregisterHandler deletes the server handler
+func (d *DNS) DeregisterHandler(_ context.Context, router *api.RoutingTree) {
+	router.Remove(http.MethodGet, route)
+}
+
+// Handler defines the server handler for the dns check
+func (d *DNS) Handler(w http.ResponseWriter, _ *http.Request) {
+	w.WriteHeader(http.StatusOK)
+}
+
+func (d *DNS) GetMetricCollectors() []prometheus.Collector {
+	return d.metrics.GetCollectors()
+}
+
+// check performs DNS checks for all configured targets using a custom net.Resolver.
+// Returns a map where each target is associated with its DNS check Result.
+func (d *DNS) check(ctx context.Context) map[string]Result {
+	log := logger.FromContext(ctx)
+	log.Debug("Checking dns")
+	if len(d.config.Targets) == 0 {
+		log.Debug("No targets defined")
+		return map[string]Result{}
+	}
+
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	results := map[string]Result{}
+
+	d.client.SetDialer(&net.Dialer{
+		Timeout: d.config.Timeout,
+	})
+
+	log.Debug("Getting dns status for each target in separate routine", "amount", len(d.config.Targets))
+	for _, t := range d.config.Targets {
+		target := t
+		wg.Add(1)
+		lo := log.With("target", target)
+
+		getDNSRetry := helper.Retry(func(ctx context.Context) error {
+			res, err := getDNS(ctx, d.client, target)
+			mu.Lock()
+			defer mu.Unlock()
+			results[target] = res
+			if err != nil {
+				return err
+			}
+			return nil
+		}, d.config.Retry)
+
+		go func() {
+			defer wg.Done()
+			status := 1
+
+			lo.Debug("Starting retry routine to get dns status")
+			if err := getDNSRetry(ctx); err != nil {
+				status = 0
+				lo.Warn("Error while looking up address", "error", err)
+			}
+			lo.Debug("Successfully got dns status of target")
```

Comment:
*Successfully* is misleading here, as it will appear after the "Error while looking up address" warning log. I would suggest replacing with a more neutral phrasing, like: "DNS check completed for target".